inductor: don't force convert channels last if one op's user is as_strided ops

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105111


For ```aten.unfold```, it will be decomposed to ```aten.as_satrided```, and it depends on input size and stride, we shouldn't change its' stride to avoid getting a wrong value.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

XiaobingSuper committed:
inductor: don't force convert channels last if one op's user is as_strided ops

[ghstack-poisoned]

XiaobingSuper commented on torch/_inductor/graph.py:
```diff
@@ -673,28 +681,26 @@ def run_node(self, n: torch.fx.Node):
-                        and n in self.nodes_prefer_channels_last
-                        and n.name not in self.user_visible_outputs
-                    ):
-                        stride_order = ir.NHWC_STRIDE_ORDER
```
@shunting314, I don't know why converting the result's stride only for one of node's users is output op or as_strided_op, I think we shouldn't have such a condition, please correct me.

shunting314 replied:
I think the point here is, in general we don't need force any stride for a node unless we return it. This way, the compiler can decide whatever stride is best it thinks when handling that node

XiaobingSuper replied:
I am confused about why converting a node results to channels last here(it should be out of this condition), there has a condition that one of the node users is graph output rather than return this node.

eellison reviewed (commented):
nice! change seems good to me, will defer to @shunting314 on this one though.

XiaobingSuper commented:
> I'm still trying to understand why the stride of inputs for `as_strided` can not change. I tried some example
> 
> ```
> >>> s = torch.rand(2, 3, 5, 5)
> >>> t = s.to(memory_format=torch.channels_last)
> >>> s.unfold(2, 2, 2).equal(t.unfold(2, 2, 2))
> True
> ```
> 
> It shows changing stride here is fine. So I'm wondering what's the correct way to trigger the problem?

I think the root cause is the decomposition of ```aten.unfold```, it can be fixed by https://github.com/pytorch/pytorch/pull/105165, but I don't know whether still other ops can be decomposed to ```as_strided```, and make the ```as_strided```inputs depend on input size and stride info.

XiaobingSuper commented:
@shunting314 , see the decomposition of ```aten.unfold``` and ```aten.unfold_copy``: https://github.com/pytorch/pytorch/blob/main/torch/_refs/__init__.py#L4059, there has an assumption that the input size and stride should not be changed. I also thought that it may be better to disable those decompositions， but it only can solve  ```unfold``` ops, there may have other ops that have similar behavior.

eellison commented:
> I'm still trying to understand why the stride of inputs for as_strided can not change. I tried some example

@shunting314 it's because as_strided is not compositional. See: [here](https://fb.workplace.com/groups/401294880339143/posts/1582738638861422/). 

For oss folks:

Why is torch.as_strided problematic?
torch.as_strided is conceptually simple but its behavior depends not only on the value of a tensor, but also its representation. If we chain two calls to torch.as_strided, the second call completely ignores the first call.
We are concerned about this when the output of a previous operation yields a tensor of uncertain stride. For example, with [torch.reshape](https://l.workplace.com/l.php?u=https%3A%2F%2Fpytorch.org%2Fdocs%2Fstable%2Fgenerated%2Ftorch.reshape.html&h=AT1U--ko4o6bGOuu4ns9FVTz26LdGD5C7ylQeNhGt1vXPTNtL6l04AMpIk4ZuHdEV5swrsHQMRdIY9WEwf2ts9AmNCdNJ3xVSxIgbr94u5JcioA24suQWbwciYQMUSw8y4OiC3avSshnIDGQ0j3IR6fUcWZ2b3BBLT_jW8R7N2EDl1MqFg&__tn__=-UK-R&c[0]=AT2wMfONGNBYIa9X_UrL1D5oblME3k706Fb3lRS4yjr3tYtxnM_lLUqdi2R2lZZWYh--c5pDlehJVbrpOpZCiHSY7MPiE7w2HQ88uUJqHhHjZVQ05-IShnz_B2JicihcrZCnkE5zGxV6O60RChGKitFWSMGrkw0o3AbuzqAtf-IuaNJGjYt3k5E), it’s an operator implementation detail whether the output is a view of the input or a copy. A decomposition using the result of this operator would need to know those implementation details to understand what to do with a subsequent torch.as_strided call.
An illustrative example
```
>>> x = torch.arange(16)
>>> x
tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])
>>> xp = x.as_strided(size=[8], stride=[2])
>>> xp
tensor([ 0,  2,  4,  6,  8, 10, 12, 14])
>>> xpp = xp.as_strided(size=[4], stride=[2])
>>> xpp
tensor([0, 2, 4, 6])
```
Note that the stride of xpp (x") is 2 rather than 4, which is what we would expect if torch.as_strided composed.
Compositional as_strided
Compositional torch.as_strided, code name torch.strided_view, solves this problem by applying the strides to the value of the tensor, rather than to the storage. Proposed interface:
```
>>> x = torch.arange(16)
>>> x
tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])
>>> xp = x.strided_view(size=[8], stride=[2])
>>> xp
tensor([ 0,  2,  4,  6,  8, 10, 12, 14])
>>> xpp = xp.strided_view(size=[4], stride=[2])
>>> xpp
tensor([0, 4, 8, 12])
```

shunting314 commented:
> I think the root cause is the decomposition of aten.unfold, it can be fixed by https://github.com/pytorch/pytorch/pull/105165, but I don't know whether still have other ops that can be decomposed to as_strided, and make the as_stridedinputs depend on input size and stride info.

Yea, indeed the root cause is the decomposition rule. The rule itself is correct since it follows how eager implements `unfold`. But unfortunately, apply the rule will burn in the eager strides of input to the graph. Later on if the compiler decides a different stride for the input, it will be inconsistent with the burned in stride.

I think we should remove the decomposition rule and do lowering in inductor instead. That can make sure we don't burn in eager time stride into the graph. We should try to keep stride as implementation detail and something compiler can decide ( in most cases. There are exceptions though). If there are other ops having similar behavior as unfold, we can do similar thing to them.

By doing that, I was hoping we can remove the special handling for as_strided here: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/graph.py#L683  . We handle as_strided specially not because as_strided itself, but because of the upstream op (`unfold`) that get decomposed into as_strided (at least in this PR). It's a bit counter-intuitive to understand the motivation.

I also want to learn more from you @XiaobingSuper  if there is any other blocker for removing that special handling.

That being said, can we rely on https://github.com/pytorch/pytorch/pull/105165  + 'removing the decomposition rule for unfold' to fix the issue here?

shunting314 commented:
@peterbell10  Agreed. We still need the special handling of as_strided in GraphLowering code even if we use lowering rather than decomposing to handle unfold.

@XiaobingSuper  I think the implementation in this PR will have some perf impact since now activations which can be channels-last will be converted to eager stride (contiguously most likely). Give me some time to think about how to avoid that. I'll give you some update today.

shunting314 commented:
A possible fix that should not have much perf impact:

```
diff --git a/torch/_inductor/graph.py b/torch/_inductor/graph.py
index 7f529abfe0f..362915fb930 100644
--- a/torch/_inductor/graph.py
+++ b/torch/_inductor/graph.py
@@ -679,9 +679,9 @@ class GraphLowering(torch.fx.Interpreter):
                 torch.ops.aten.as_strided_.default,
                 torch.ops.aten.as_strided_scatter.default,
             ]
-            if any(
-                user.op == "output" or user.target in as_strided_ops for user in n.users
-            ) and isinstance(n.meta["val"], torch.Tensor):
+            is_output = any(user.op == "output" for user in n.users)
+            is_input_for_as_strided = any(user.target in as_strided_ops for user in n.users)
+            if (is_output or is_input_for_as_strided) and isinstance(n.meta["val"], torch.Tensor):
                 strides = n.meta["val"].stride()
                 dense = torch._prims_common.is_non_overlapping_and_dense(n.meta["val"])
                 # requiring a stride order for a non-dense output wouldn't
@@ -693,6 +693,7 @@ class GraphLowering(torch.fx.Interpreter):
                         len(result.get_size()) == 4
                         and n in self.nodes_prefer_channels_last
                         and n.name not in self.user_visible_outputs
+                        and not is_input_for_as_strided
                     ):
                         stride_order = ir.NHWC_STRIDE_ORDER
```

XiaobingSuper committed:
Update on "inductor: don't force convert channels last if one op's user is as_strided ops"



For ```aten.unfold```, it will be decomposed to ```aten.as_satrided```, and it depends on input size and stride, we shouldn't change its' stride to avoid getting a wrong value.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: don't force convert channels last if one op's user is as_strided ops"



For ```aten.unfold```, it will be decomposed to ```aten.as_satrided```, and it depends on input size and stride, we shouldn't change its' stride to avoid getting a wrong value.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

XiaobingSuper commented:
> A possible fix that should not have much perf impact:
> 
> ```
> diff --git a/torch/_inductor/graph.py b/torch/_inductor/graph.py
> index 7f529abfe0f..362915fb930 100644
> --- a/torch/_inductor/graph.py
> +++ b/torch/_inductor/graph.py
> @@ -679,9 +679,9 @@ class GraphLowering(torch.fx.Interpreter):
>                  torch.ops.aten.as_strided_.default,
>                  torch.ops.aten.as_strided_scatter.default,
>              ]
> -            if any(
> -                user.op == "output" or user.target in as_strided_ops for user in n.users
> -            ) and isinstance(n.meta["val"], torch.Tensor):
> +            is_output = any(user.op == "output" for user in n.users)
> +            is_input_for_as_strided = any(user.target in as_strided_ops for user in n.users)
> +            if (is_output or is_input_for_as_strided) and isinstance(n.meta["val"], torch.Tensor):
>                  strides = n.meta["val"].stride()
>                  dense = torch._prims_common.is_non_overlapping_and_dense(n.meta["val"])
>                  # requiring a stride order for a non-dense output wouldn't
> @@ -693,6 +693,7 @@ class GraphLowering(torch.fx.Interpreter):
>                          len(result.get_size()) == 4
>                          and n in self.nodes_prefer_channels_last
>                          and n.name not in self.user_visible_outputs
> +                        and not is_input_for_as_strided
>                      ):
>                          stride_order = ir.NHWC_STRIDE_ORDER
> ```

@shunting314, thanks for your patch, please help review PR again.

XiaobingSuper committed:
Update on "inductor: don't force convert channels last if one op's user is as_strided ops"



For ```aten.unfold```, it will be decomposed to ```aten.as_satrided```, and it depends on input size and stride, we shouldn't change its' stride to avoid getting a wrong value.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge
